domain/storageprovisioning/service/entitylife.go

```diff
@@ -0,0 +1,136 @@
+// Copyright 2025 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package service
+
+import (
+	"context"
+	"maps"
+	"slices"
+
+	"github.com/juju/juju/core/changestream"
+	"github.com/juju/juju/core/database"
+	"github.com/juju/juju/core/watcher/eventsource"
+	"github.com/juju/juju/domain/life"
+	"github.com/juju/juju/internal/errors"
+)
+
+// EntityLifeGetter is a function which returns the current life for each
+// concerned entity involved in the context. The string id value used for
+// entities must be consistent across multiple calls.
+//
+// The purpose of this function is to help work out what storage entities that
+// a machine provisioner cares about have had a life change.
+type EntityLifeGetter func(context.Context) (map[string]life.Life, error)
+
+// EntityLifeMapperFunc provides a watcher mapper that can be used to
+// take change events for one concern and translate this into a set of entity
+// life changes. Entity in this case is a storage entity that is associated
+// with the context of the current cocern.
+//
+// A concrete example would be a machine provisioner concern where the concern
+// is the net node of the machine being watched. Entities are all the storage
+// entities that would be provisioned from this net node context.
+//
+// When an entities life has change compared with the last seen value or an
+// entity has been added or removed it will result in the id being returned in
+// the change set.
+//
+// [EntityLifeGetter] provides the latest life values for the concerns entities.
+// This func expect to be provided with a func that provides the initial life
+// values to start performing change detection on top of. This func will be
+// called once on the first invocation of the returned mapper.
+//
+// The returned mapper is not thread safe.
+func EntityLifeMapperFunc(
+	initialLife func(ctx context.Context) (map[string]life.Life, error),
+	lifeGetter EntityLifeGetter,
+) eventsource.Mapper {
+	var haveInitialLife bool
+	var knownLife map[string]life.Life
+	return func(
+		ctx context.Context, _ []changestream.ChangeEvent,
+	) ([]string, error) {
+		if !haveInitialLife {
```

Comment:
You can simply check if knownLife == nil directly, no need to define a separate boolean variable.
Alternatively, you could retrieve the initial life earlier, before the return.

Reply:
what happens if the initial query returns nil for the initial life? One problematic implementation could make this fail in subtle ways.